solidity/nft_token.sol

```diff
@@ -0,0 +1,254 @@
+
+                // Check the PDA header data
+                uint64 header = ai.data.readUint64LE(0);
+                if (header != pdaHeader) {
+                    revert("Wrong pda header");
+                }
+
+                // Check the PDA address correctness
+                (address pdaPosition, ) = try_find_program_address(["position", position.position_mint], orca);
+                // TODO: Uncomment following lines in production
+//                if (pdaPosition != positionDataAccount) {
+//                    revert("Wrong position pdaPosition");
+//                }
+
+                return position;
+            }
+        }
+
+        revert("account missing");
+    }
+
+    @mutableAccount(fromTokenAccount)
+    @mutableAccount(pdaTokenAccount)
+    @mutableAccount(toErc20)
+    @mutableAccount(mintERC20)
+    @account(positionDataAccount)
+    @signer(fromWallet)
+    function deposit(address positionMint) external {
+        AccountInfo ai = tx.accounts.positionDataAccount;
+        // TODO Shift everything left by 8 bytes in production
+        Position position = Position({
+            whirlpool: ai.data.readAddress(16),
+            position_mint: ai.data.readAddress(48),
+            liquidity: ai.data.readUint128LE(80),
+            tick_lower_index: ai.data.readInt32LE(96),
+            tick_upper_index: ai.data.readInt32LE(100)
+        });
+
+        // Check the whirlpool
+        if (position.whirlpool != whirlpool) {
+            revert("Wrong whirlpool address");
+        }
+
+        // Check the NFT address
+        if (position.position_mint != positionMint) {
+            revert("Wrong NFT address");
+        }
+
+        // Check tick values
+        if (position.tick_lower_index != min_tick_lower_index || position.tick_upper_index != max_tick_lower_index) {
+            revert("Wrong ticks");
+        }
+
+        // Check the PDA ownership
+        // TODO: Uncomment following lines in production
+//        if (ai.owner != orca) {
+//            revert("Wrong pda owner");
+//        }
+
+        // Check the PDA header data
+        uint64 header = ai.data.readUint64LE(0);
+        if (header != pdaHeader) {
+            revert("Wrong pda header");
+        }
+
+        // Check the PDA address correctness
+        (address pdaPosition, ) = try_find_program_address(["position", position.position_mint], orca);
+        // TODO: Uncomment following lines in production
+//        if (pdaPosition != positionDataAccount) {
+//            revert("Wrong position pda");
+//        }
+
+        // TODO: Do the liquidity check for max(uint64) value as it is provided as uint128 from the LP provider
+        uint64 liquidity = uint64(position.liquidity);
+
+        // Transfer the NFT to the pdaTokenAccount address of this program
+        SplToken.transfer(
+            tx.accounts.fromTokenAccount.key,
+            tx.accounts.pdaTokenAccount.key,
+            tx.accounts.fromWallet.key,
+            1);
+
```

Comment:
Not sure if reentrancy is a thing in Solana, still suggest doing checks-effects-interactions pattern here and elsewhere! It's a good habit imo

Reply:
Yes. for solana reentrancy issue very rare.
1) https://medium.com/coinmonks/on-smart-contracts-why-solana-is-more-secure-18626b09929b
- CPI is limited, we can endlessly call one contract from another into the EVM until the gas runs out. in Solana CPI - 4 max.
- there is no concept of a token as a program. there are no arbitrary program (contract) made by users. we call a limited number of trusted programs.
For us, one of the system programs (like SPL-token) or trusted programs (like orca) must be malicious for potencial reentrancy. the end-user provides not programs but accounts.